### ISSUE TITLE
Saving Uint8Array seems to take way too much space?

### ISSUE THREAD

**Issue by sid-kap** [dfahlander/Dexie.js#665]
I'm trying to use Dexie to store an object that contains 2-3 images, each 2-20 MB, in the format of Uint8Arrays.
I expect that my object should be around 50-60MB total, but when I try to store it, I get the error:
```
Unhandled rejection: UnknownError: The serialized value is too large (size=268436288 bytes, max=267386880 bytes).
```
Which suggests that the object is about 280 MB.

I have a few theories for why this might be happening. Maybe one of these is the reason?
* Is there some space overhead while storing Uint8Array in Dexie/IndexedDB? Would I get any space savings by using Blob instead?
* Also, my object is highly nested (each image is stored as an object that contains num rows, num cols, and the Uint8Array). Could this nesting be causing the problem? (i.e., maybe it's stringifying the Uint8Arrays instead of saving them as binary objects?)

**Comment by dfahlander** [comment]
No, we're basically storing the object as is. There's no special handling of Blobs or Uint8Arrays. Some browsers may have bugs though. What browser are you using?

**Comment by sid-kap** [comment]
Firefox 57.0.3

**Comment by sid-kap** [comment]
Yep, looks like the big `ArrayBuffer` was the problem. Here's an example to reproduce:

```typescript
import Dexie from "dexie"

import "./index.html"

class MyDatabase extends Dexie {
    myTable: Dexie.Table<State, number>

    constructor () {
        super("MyDatabase")
        this.version(1).stores({
            myTable: '++id, array',
        })
    }
}

interface State {
    id?: number,
    array: Uint8Array,
}

console.log("script running")
const db = new MyDatabase()

main()

async function main() {
    console.log("in main")
    const smallArray = new Uint8Array([1,2,3])
    const result = await db.myTable.put({id: 1, array: smallArray})
    console.log("First insert was successful with result", result)

    const buffer = new ArrayBuffer(3e8)
    const bigArray = new Uint8Array(buffer, 0, 3)
    const result2 = await db.myTable.put({id: 2, array: bigArray})
    console.log("Second insert was successful with result", result2)
}
```

The output is:
```
First insert was successful with result 1
Unhandled rejection: UnknownError: The serialized value is too large (size=300000098 bytes, max=267386880 bytes).
```
So it succeeds on the first one, but fails with the "too large" error on the second one. Looks like this is a Firefox bug?

**Comment by dfahlander** [comment]
I would suppose so, unless the the issue lies in the spec. Would be nice just to check if chrome has the same behavior. If not, I would post it on Bugzilla. They're quite fast with fixing these kinds of bugs and it seems to be easily fixed.

Another thing : do you intentionally want to index the array property? Seems like a heavy index if it contains an image.

**Comment by dfahlander** [comment]
@sid-kap Could this the same problem as in #667? Browser (and the IDB specification) behaves very different with indexed properties than non-indexed. Basically, indexes on binary keys will index the backing array buffer. So it may not be a bug. The issue may be that you are using a very large index. And if it is an image, there's no reason to index it.

**Comment by sid-kap** [comment]
You're right, I was indexing the typed array. When I get a chance I'll try it again without indexing that field and I expect that it should work.

I think the documentation of this could be improved. I I don't remember reading that I should only pass in the indexed fields into Version.stores

**Comment by sid-kap** [comment]
Huh, removing the index doesn't seem to fix it. Do I need to "reset" my database somehow to get rid of the index?

**Comment by sid-kap** [comment]
Yeah, even incrementing the version and deleting the database didn't fix it. I submitted an issue to Firefox, hopefully they will get back soon.
In the mean time, I am using `ArrayBuffer.prototype.slice` to move my data to a smaller ArrayBuffer, which is working.

**Comment by sid-kap** [comment]
Also, looks like it works in Chrome, although it is very slow (took less than a second for the first insert, and took like 20 seconds for the second one). So must be a bug there as well.

**Comment by sid-kap** [comment]
Thanks for updating the homepage. In the comment that says "Define a schema", can you instead change it to "Define the indexed columns in the schema" or something along those lines?

I think warning about storing large indexes would be a good idea. I don't think slicing large arrays is necessary though, since that should hopefully be fixed by the browsers at some point.
